### PR TITLE
Add check_case_conflict hook

### PR DESCRIPTION
**Pull request by guykisel**
This is for #21. I'm not 100% sure about the validity of my approach or of my test cases, so comments/criticism are welcome.

### CODE REVIEW

**Comment by guykisel** [comment]
Hmm, the tests failed in Travis when I tried to commit to the temporary git repo. Not really sure what to do about that.

File: pre_commit_hooks/check_case_conflict.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+    # Compare the files pre-commit says are modified against all files in the
+    # repo
+    pairs = itertools.product(filenames, repo_filenames)
+
+    retv = 0
+    for pair in pairs:
+        if pair[0] != pair[1] and pair[0].lower() == pair[1].lower():
+            print('{0} and {1} differ only in case.'.format(pair[0], pair[1]))
+            retv = 1
```
Comment (asottile) [inline pre_commit_hooks/check_case_conflict.py]: a more efficient (and maybe more correct?) approach:

```
1. repo_files = set(`git ls-files`)
2. added_files = set((I forget, find the code in the large files thing, make it into a util that both use))
3. pre_commit_files = set((get these from argv))

relevant_files = pre_commit_files & added_files
repo_files -= relevant_files

if set(x.lower() for x in repo_files) & set(x.lower() for x in relevant_files):
    # print useful error message

```

Comment (asottile) [inline pre_commit_hooks/check_case_conflict.py]: oh you'll also want to check `len(relevant_files) == len(set(x.lower() for x in relevant_Files))`

File: pre_commit_hooks/check_case_conflict.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+    return retv
+
+
+def main(argv=None):
+    argv = argv if argv is not None else sys.argv[1:]
```
Comment (asottile) [inline pre_commit_hooks/check_case_conflict.py]: so I've found that parser.parse_args(None) will magically use sys.argv, so you can actually drop this line

**Comment by asottile** [comment]
as for travis, we solved this here: https://github.com/pre-commit/pre-commit/blob/master/.travis.yml#L10

**Comment by guykisel** [comment]
Cool, thanks for the suggestions!

**Comment by guykisel** [comment]
Tests are passing after the tox config fix, but there are errors due to duplicate code. I'm working on those errors + the suggestions above.

**Comment by guykisel** [comment]
So while I'm working on this, I've realized that the unit tests for this mostly only work on a case-sensitive filesystem. Does it seem like a bad idea to do this:
```python
def filesystem_is_case_insensitive():
    """From http://stackoverflow.com/a/7870076/2047682"""
    # By default mkstemp() creates a file with
    # a name that begins with 'tmp' (lowercase)
    tmphandle, tmppath = tempfile.mkstemp(dir=os.getcwd())
    case_insensitive = os.path.exists(tmppath.upper())
    os.remove(tmppath)
    return case_insensitive

def test_adding_something_with_conflict(temp_git_dir):
    if filesystem_is_case_insensitive():
        return
    with local.cwd(temp_git_dir):
        write_file('f.py', "print('hello world')")
        local['git']('add', 'f.py')
        write_file('F.py', "print('hello world')")
        local['git']('add', 'F.py')

        assert find_conflicting_filenames(['f.py', 'F.py']) == 1
```

**Comment by guykisel** [comment]
Also, weird note about developing this on OS X (which defaults to case-insensitive) - I created a case-sensitive disk image to work in, then had to override pytest's default base temporary directory, because it was creating temporary files in the case-insensitive volume.

**Comment by guykisel** [comment]
Note this is still a work in progress. My conflict detection still seems overly complicated.

**Comment by guykisel** [comment]
On my macbook, the current algorithm takes ~40ms for 10k filenames, and ~1s for 1m filenames.

File: tox.ini [hunk 3]
```diff
@@ -8,7 +8,7 @@ install_command = pip install --use-wheel {opts} {packages}
 deps = -rrequirements-dev.txt
 commands =
     coverage erase
-    coverage run -m pytest {posargs:tests}
+    coverage run -m pytest --basetemp=tmp {posargs:tests}
```
Comment (asottile) [inline tox.ini]: You can accomplish this in userspace by doing:

`TMPDIR=tmp tox -e py27` (or `TMPDIR=tmp make test`)

File: pre_commit_hooks/check_case_conflict.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+def find_conflicting_filenames(filenames):
+    repo_files = set(local['git']('ls-files').splitlines())
+    relevant_files = set(filenames) | added_files()
+    repo_files -= relevant_files
+    retv = 0
+
+    # new file conflicts with existing file
+    conflicts = (set(x.lower() for x in repo_files)
+                 & set(x.lower() for x in relevant_files))
```
Comment (asottile) [inline pre_commit_hooks/check_case_conflict.py]: just a nit:

I usually prefer this written as:

conficts = (
    (...) &
    (...)
)

File: pre_commit_hooks/check_case_conflict.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    lowercase_relevant_files = set(x.lower() for x in relevant_files)
+    for filename in set(relevant_files):
+        if filename.lower() in lowercase_relevant_files:
+            lowercase_relevant_files.remove(filename.lower())
+        else:
+            conflicts.add(filename.lower())
+
+    if conflicts:
+        conflicting_files = [x for x in repo_files | relevant_files if
+                             x.lower() in conflicts]
```
Comment (asottile) [inline pre_commit_hooks/check_case_conflict.py]: another nit:

I'd write this as:
```
conflicting_files = [
     x for x in repo_files | relevant_files
    if x.lower() in conflicts
]
```

File: pre_commit_hooks/check_case_conflict.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+
+def find_conflicting_filenames(filenames):
+    repo_files = set(local['git']('ls-files').splitlines())
+    relevant_files = set(filenames) | added_files()
+    repo_files -= relevant_files
+    retv = 0
+
+    # new file conflicts with existing file
+    conflicts = (set(x.lower() for x in repo_files)
```
Comment (asottile) [inline pre_commit_hooks/check_case_conflict.py]: seems like this code might be slightly more readable with a

```
def lower_set(iterable):
    return set(x.lower() for x in iterable)
```

**Comment by asottile** [comment]
Overall seems fine.

- To fix the pylint stuff, move the `temp_git_dir` fixture into tests/conftest.py
- Remove the `--basetemp`
- Squash commits
- Might need to add this to pylintrc also:
```
[SIMILARITIES]
# maybe min-similarity-lines=N as well?
ignore-imports=yes
```
And I'll merge this :D

Thanks for the PR!

**Comment by guykisel** [comment]
Fixed and squashed.

**Comment by asottile** [comment]
@guykisel Thanks for the PR :D

**Comment by guykisel** [comment]
:)

**Comment by asottile** [comment]
Ooh one more thing, can you update the README?

**Comment by guykisel** [comment]
Oh, sure. Want a new PR for that?

**Pull request merged**